tools/crm_resource_print.c

```diff
@@ -158,41 +158,61 @@ agent_status_default(pcmk__output_t *out, va_list args) {
-                  services_ocf_exitcode_str(rc), rc);
+        /* Operation <action> [for <resource>] (<class>[:<provider>]:<agent>)
+         * returned <exit-code> (<exit-description>[: <exit-reason>])
+         */
+        out->info(out, "Operation %s%s%s (%s%s%s:%s) returned %d (%s%s%s)",
+                  action,
+                  ((name == NULL)? "" : " for "), ((name == NULL)? "" : name),
+                  class, (provider? ":" : ""), (provider? provider : ""), type,
+                  rc, services_ocf_exitcode_str(rc),
+                  ((exit_reason == NULL)? "" : ": "),
+                  ((exit_reason == NULL)? "" : exit_reason));
     } else {
-        out->err(out, "Operation %s%s%s (%s%s%s:%s) failed: '%s' (%d)",
-                 action, name ? " for " : "", name ? name : "",
-                 class, provider ? ":" : "", provider ? provider : "", type,
-                 pcmk_exec_status_str(status), status);
+        /* Operation <action> [for <resource>] (<class>[:<provider>]:<agent>)
+         * could not be executed (<execution-status>[: <exit-reason>])
+         */
+        out->err(out,
+                 "Operation %s%s%s (%s%s%s:%s) could not be executed (%s%s%s)",
+                 action,
+                 ((name == NULL)? "" : " for "), ((name == NULL)? "" : name),
+                 class, (provider? ":" : ""), (provider? provider : ""), type,
+                 pcmk_exec_status_str(status),
+                 ((exit_reason == NULL)? "" : ": "),
+                 ((exit_reason == NULL)? "" : exit_reason));
     }
 
     return pcmk_rc_ok;
 }
 
 PCMK__OUTPUT_ARGS("agent-status", "int", "const char *", "const char *", "const char *",
-                  "const char *", "const char *", "int")
+                  "const char *", "const char *", "int", "const char *")
 static int
 agent_status_xml(pcmk__output_t *out, va_list args) {
-    int status G_GNUC_UNUSED = va_arg(args, int);
+    int status = va_arg(args, int);
     const char *action G_GNUC_UNUSED = va_arg(args, const char *);
     const char *name G_GNUC_UNUSED = va_arg(args, const char *);
     const char *class G_GNUC_UNUSED = va_arg(args, const char *);
     const char *provider G_GNUC_UNUSED = va_arg(args, const char *);
     const char *type G_GNUC_UNUSED = va_arg(args, const char *);
     int rc = va_arg(args, int);
+    const char *exit_reason = va_arg(args, const char *);
 
-    char *status_str = pcmk__itoa(rc);
+    char *exit_str = pcmk__itoa(rc);
+    char *status_str = pcmk__itoa(status);
 
     pcmk__output_create_xml_node(out, "agent-status",
-                                 "code", status_str,
+                                 "code", exit_str,
                                  "message", services_ocf_exitcode_str(rc),
+                                 "execution_code", status_str,
+                                 "execution_message", pcmk_exec_status_str(status),
+                                 "reason", (exit_reason? exit_reason : ""),
```

Comment:
If there's no exit_reason, I think I'd prefer to see this attribute left off the agent-status element rather than have its value be an empty string.

Reply:
Me too, but how? pcmk__output_create_xml_node() -> pcmk__xe_set_propv() aborts on NULLs. I suppose we could make it skip NULL values instead

Reply:
It uses NULL as the argument terminator. However it should be reasonable to use NULL name as the argument terminator, and skip NULL values -- it's internal API and that's how it's currently used anyway

Reply:
Ah, I didn't think of that. What do you think of the fix here (the "skip  NULL values" commit) skipping NULL values in pcmk__output_create_xml_node()? I don't see any potential problems with it

Reply:
I think that's a fine approach, and it lets us get rid of all the places I'm doing the other method (grep says there's 34 that need to be investigated).